### PR TITLE
[cxx-interop] Add support for reference counting operations on foreign reference types.

### PR DESCRIPTION
**Pull request by zoecarver**
Based on #59439

### CODE REVIEW

**Comment by zoecarver** [comment]
This is still a bit of a wip: I need to change a lot of these asserts into diagnostics.

**Comment by hyp** [comment]
Could you also add an execution test that covers the use of the refcounted FRT in a generic context?

Also, perhaps a struct that has stored refcounted FRT in a generic context as well, that would be nice to validate.

**Review by rjmccall — CHANGES_REQUESTED** [review]

File: lib/IRGen/GenMeta.cpp [hunk 1]
```diff
@@ -5532,6 +5518,60 @@ namespace {
...
+      // ourselves the load-time work.
+      B.addNullPointer(IGM.TypeMetadataPtrTy);
+
+      assert(!Target->hasSuperclass() &&
+             "foreign reference types should not have superclasses.");
+    }
+
+    void addReservedWord() {
+      B.addNullPointer(IGM.Int8PtrTy);
+    }
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/IRGen/GenMeta.cpp:5573]: This PR was out of date. I fixed this in the previous PR that this one is based on. Once I rebase the changes will be reflected.

File: lib/IRGen/GenHeap.cpp [hunk 2]
```diff
@@ -1015,6 +1022,8 @@ void IRGenFunction::emitStrongRelease(llvm::Value *value,
     return emitBridgeStrongRelease(value, atomicity);
   case ReferenceCounting::Error:
     return emitErrorStrongRelease(value);
+  case ReferenceCounting::CxxCustom:
+    llvm_unreachable("Ref counting should be handled in TypeInfo.");
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply lib/IRGen/GenHeap.cpp:1026]: That's what I would have liked to do, but that would require trafficking an optional ClassDecl to this case. While not ideal, I think my current solution is the best option without doing a major refactor. If you'd prefer me to traffic an optional ClassDecl through, I'm happy to do that instead.

File: stdlib/public/runtime/Casting.cpp [hunk 3]
```diff
@@ -828,6 +841,11 @@ swift_dynamicCastUnknownClassUnconditionalImpl(const void *object,
 #endif
   }
 
+  // Foreign reference types don't suppport casting.
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply stdlib/public/runtime/Casting.cpp:844]: Good point. I updated this comment to be more correct and added some radar links.

File: include/swift/AST/ReferenceCounting.h [hunk 4]
```diff
@@ -33,6 +33,10 @@ enum class ReferenceCounting : uint8_t {
   /// types.
   None,
 
+  /// The object uses swift_attr("retain:XXX") and "release:XXX" to implement
+  /// reference counting.
+  CxxCustom,
```
[the remark this answers is not in the document]
Comment (zoecarver) [reply include/swift/AST/ReferenceCounting.h:38]: Good call. Done.

**Comment by zoecarver** [comment]
> This is still a bit of a wip: I need to change a lot of these asserts into diagnostics.

I think it may be better to do this after the fact. Just so that we can break this PR up a bit and make it easier to review. Then we can have a PR to bike shed the diagnostics. WDYT?

**Comment by zoecarver** [comment]
@swift-ci please smoke test

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux

**Comment by zoecarver** [comment]
@swift-ci please test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please test.

**Pull request merged**